doc/reference/trd-appid.md

```diff
@@ -0,0 +1,396 @@
+of the public key in the header.
+
+The `Rsa3072KeyWithID` value has a data of length of 768 bytes. It contains a public 3072-bit
+RSA key (384 bytes), followed by a 384-byte ciphertext block, consisting of the SHA512 
+hash of the application binary in this Tock binary followed by a 32-bit application
+ID, encrypted by the private key of the public key in the header.
+
+The `Rsa4096KeyWithID` value has a data of length of 1024 bytes. It contains a public 4096-bit
+RSA key (512 bytes), followed by a 512-byte ciphertext block, consisting of the SHA512 
+hash of the application binary in this Tock binary followed by a 32-bit application
+ID, encrypted by the private key of the public key in the header.
+
+
+4 `Verifier` trait
+===============================
+
+The `Verifier` trait defines an interface to a module that accepts,
+passes on, or rejects application credentials. When a Tock board
+asks the kernel to load processes, it passes a reference to a 
+`Verifier`, which the kernel uses to check credentials.
+
+
+```rust
+pub enum VerificationResult {
+  Accept,
+  Pass,
+  Reject
+}
+
+pub trait Verifier {
+  fn require_credential(&self) -> bool;
+  fn check_credentials(&self, 
+                       credentials: &TbfHeaderV2Credentials, 
+                       binary: &mut [u8]) -> VerificationResult;
+}
+```
+
+The kernel, when it loads a Tock binary, scans its headers in order
+from the beginning of the Tock binary. At each
+`TbfHeaderV2Credentials` header it encounters, it calls
+`check_credentials` on the provided `Verifier`. If the `Verifier`
+returns `Accept`, the kernel stops processing credentials and
+continues loading the Tock binary. If the `Verifier` returns `Reject`,
+the kernel stops processing credentials and terminates loading the
+Tock binary. If the `Verifier` returns `Pass`, the kernel tries the
+next `TbfHeaderV2Credentials`, if there is one. 
+
+If the kernel reaches the end of the TBF headers without encountering
+a `Reject` or `Accept` result, it calls `require_credentials` to ask
+the `Verifier` what the default behavior is.  If `require_credentials`
+returns `true`, the kernel rejects the Tock binary and terminates
+loading it. If `require_credentials` returns `false`, the kernel
+accepts the Tock binary and continues loading it. If a Tock binary has
+no `TbfHeaderV2Credentials` headers then there will be no `Accept` or
+`Reject` results and `require_credentials` defines whether to load
+such a binary.
+
+An implementer of `Verifier` sets the security policy of Tock binary
+loading by deciding which types of credentials, and which credentials,
+are acceptable and which are rejected.
+
+If `check_credentials` returns `Accept` for a
+`TbfHeaderV2Credentials`, the kernel stores a reference to this
+`TbfHeaderV2Credentials` in the process structure. This data
+represents the acting credentials of the process.
+
+
+5 Short IDs and the `Compress` trait
+===============================
```

Comment:
What about short IDs for nonvolatile storage? Should they be called "storage IDs"?

Reply:
I wasn't quite ready to open that can of worms, although I'm open to it being part of this document eventually?

Reply:
I'm concerned that using the term short IDs to refer only to this particular form of shortened IDs (IDs representing application IDs the kernel "knows about") will become confusing when storage IDs are added in the future.

Reply:
Yeah, I agree, this could be confusing.

Reply:
The IDs for storage are already called `write_id`, `read_ids` and `access_ids`, so I don't think `ShortIDs` conflicts here

Reply:
Would it be sufficient to say that Short IDs are distinct from IDs used for storage, and describe their differences? 

One issue that might come up is whether storage IDs are determined by a TBF header or by a storage access controller within the kernel; e.g., the use case might want to connect storage access to Credentials, and having it in a TBF header pushes this connection to the signer. 

So we want to keep the TBF header because it's needed for some use cases, but do we also want an API to support kernel generation of storage access based on Credentials?

Reply:
> Would it be sufficient to say that Short IDs are distinct from IDs used for storage, and describe their differences?

Yes, I think they are two different things, for different goals and don't overlap. I'm not even sure if storage IDs really need to be mentioned in this TRD.

> One issue that might come up is whether storage IDs are determined by a TBF header or by a storage access controller within the kernel; e.g., the use case might want to connect storage access to Credentials, and having it in a TBF header pushes this connection to the signer.

The idea was that storage IDs are specified in the TBF header. `Verifier` can then verify that they match what it expects based on the credentials.

So for example a `Verifier` implementation could calculate or have a list of storage IDs mapping to short IDs to double check. But the actual IDs would come from the TBF. That way storage IDs are always specified in the same way, while still allowing stricter verification. Hopefully simplifying implementations

Reply:
> So for example a `Verifier` implementation could calculate or have a list of storage IDs mapping to short IDs to double check.

The only way I can see that working is if the `Verifier` has access to nonvolatile writable storage. Is that what you intended? That doesn't seem desirable to me from a complexity standpoint.

Reply:
It could be included in the Verifier function implementation, included in the signed application binary that Verifier is verifying or in non-volatile storage.